filters/filter.go

```diff
@@ -0,0 +1,485 @@
+
+The formal grammar is as follows:
+
+selectors := expr ("," expr)*
+selector  := fieldpath (operator value)
+fieldpath := field ('.' quoted)*
+field     := quoted | [A-Za-z] [A-Za-z0-9_]+
+value     := quoted | [^\s,]+
+quoted    := '"' [^"]+ (('\"' | '\\') [^"]*)* '"'
+
+*/
+func Parse(s string) (Filter, error) {
+	p := parser{s: s}
+	return p.parse()
+}
+
+type parser struct {
+	s string
+	i int
+}
+
+func (p *parser) parse() (Filter, error) {
+	ss, err := p.selectors()
+	if err != nil {
+		return nil, err
+	}
+
+	if !p.eof() {
+		return nil, p.mkerr(errInvalidFormat)
+	}
+
+	return matchers(ss), nil
+}
+
+func (p *parser) selectors() ([]matcher, error) {
+	s, err := p.selector()
+	if err != nil {
+		return nil, err
+	}
+
+	ss := []matcher{s}
+
+	if p.eof() {
+		return ss, nil
+	}
+
+	for !p.eof() {
+		if !p.maybe(",") {
+			break
+		}
+
+		s, err := p.selector()
+		if err != nil {
+			return nil, err
+		}
+
+		ss = append(ss, s)
+	}
+
+	return ss, nil
+}
+
+func (p *parser) selector() (matcher, error) {
+	fieldpath, err := p.fieldpath()
+	if err != nil {
+		return matcher{}, err
+	}
+
+	if p.eof() || !isOp(p.peek()) {
+		return matcher{
+			fieldpath: fieldpath,
+			operator:  operatorPresent,
+		}, nil
+	}
+
+	op, err := p.operator()
+	if err != nil {
+		return matcher{}, err
+	}
+
+	value, err := p.value()
+	if err != nil {
+		if err == io.EOF {
+			return matcher{}, io.ErrUnexpectedEOF
+		}
+		return matcher{}, err
+	}
+
+	return matcher{
+		fieldpath: fieldpath,
+		value:     value,
+		operator:  op,
+	}, nil
+}
+
+func (p *parser) fieldpath() ([]string, error) {
+	f, err := p.field()
+	if err != nil {
+		if err == io.EOF {
+			return nil, io.ErrUnexpectedEOF
+		}
+		return nil, err
+	}
+
+	fs := []string{f}
+	for {
+		if !p.maybe(".") {
+			break
+		}
+
+		f, err := p.quotedOrField()
+		if err != nil {
+			if err == io.EOF {
+				return nil, io.ErrUnexpectedEOF
+			}
+
+			return nil, err
+		}
+
+		fs = append(fs, f)
+	}
+
+	return fs, nil
+}
+
+func (p *parser) quotedOrField() (string, error) {
+	q, err := p.quoted()
+	if err == nil {
+		return q, nil
+	} else if err == errNotQuoted {
+		return p.field()
+	}
+
+	return "", errors.Wrap(err, "expected quoted string")
+
+}
+
+func (p *parser) field() (string, error) {
+	f, err := p.consume(isField)
+	if err != nil {
+		return "", err
+	}
+
+	return f, nil
+}
+
+func (p *parser) quoted() (string, error) {
+	if p.eof() {
+		return "", p.mkerr(io.EOF)
+	}
+	if p.peek() != '"' {
+		return "", errNotQuoted
+	}
+
+	var buffer []rune
+loop:
+	for p.advance() {
+		switch p.peek() {
+		case '"':
+			break loop
+		case '\\':
+			if !p.advance() {
+				return "", p.mkerr(io.EOF)
+			}
+			switch p.peek() {
+			case '\\':
+				buffer = append(buffer, '\\')
+			case '"':
+				buffer = append(buffer, '"')
+			default:
+				return "", p.mkerr(errors.New("illegal escape"))
+			}
+		default:
+			buffer = append(buffer, rune(p.peek()))
+		}
+	}
+
+	p.advance() // consumes final quote character
```

Comment:
Could this just be done before `break loop`. In the case where it does hit eof is double advances, while not a problem just seems unnecessary to have this here with the explaining comment.